in-progress/8401-proof-timeliness/design.md

```diff
@@ -0,0 +1,268 @@
+For example, it is highly unlikely that the rollup contract will hold the bond.
+
+The key "trick" is that we prune the pending chain as part of a proposer's block submission.
+
+An observation is that in the current design, the committee does not change in the event that no proof claim is posted for the previous epoch.
+
+```solidity
+
+contract Rollup {
+    struct ChainTip {
+        uint256 blockNumber;
+        uint256 slotNumber;
+    }
+
+    struct State {
+        ChainTip pendingTip;
+        ChainTip provenTip;
+    }
+
+    struct ProposalLog {
+        bytes32 hash;
+        uint256 slotNumber;
+        bytes32 archive;
+    }
+
+    struct ProofClaim {
+        address bondProvider;
+        address rewardRecipient;
+        uint256 claimedEpoch;
+    }
+
+    struct Fee {
+        address recipient;
+        uint256 amount;
+    }
+
+    struct EpochHeader {
+        uint256 finalBlockNumber;
+        Fee[] fees;
+    }
+
+    // State variables
+    State public state;
+    mapping(uint256 => ProposalLog) public proposalLogs;
+    ProofClaim public proofClaim;
+    IERC20 public testToken;
+    IFeeJuicePortal public feeJuicePortal;
+
+    // Constants
+    uint256 public constant CLAIM_DURATION = 13;
+    uint256 public constant EPOCH_DURATION = 32;
+    uint256 public constant PROOF_COMMITMENT_BOND_AMOUNT = 100;
+
+    // Events
+    event ProofRightClaimed(address indexed claimer, uint256 indexed epoch);
+    event PendingChainPruned(uint256 blockNumber, uint256 slotNumber);
+    event ProofSubmitted(uint256 indexed epoch, uint256 finalBlockNumber);
+
+    // Errors
+    error Rollup__NonProposerCannotClaimProofRight();
+    error Rollup__ProofRightAlreadyClaimed();
+    error Rollup__NotInClaimPhase();
+    error Rollup__NoProposalLog();
+    error Rollup__NotPreviousEpoch();
+    error Rollup__NotLastBlockInEpoch();
+    error Rollup__ProofProductionPhaseEnded();
+    error Rollup__UnauthorizedProofSubmitter();
+
+    constructor(address _testToken, address _feeJuicePortal) {
+        testToken = IERC20(_testToken);
+        feeJuicePortal = IFeeJuicePortal(_feeJuicePortal);
+    }
+
+    function claimProofRight(address _bondProvider) external {
+        uint256 currentSlot = getCurrentSlot();
+        address currentProposer = getCurrentProposer();
+
+        if (currentProposer != address(0) && currentProposer != msg.sender) {
+            revert Rollup__NonProposerCannotClaimProofRight();
+        }
+
+        if (proofClaim.rewardRecipient != address(0)) {
+            revert Rollup__ProofRightAlreadyClaimed();
+        }
+
+        if (currentSlot % EPOCH_DURATION >= CLAIM_DURATION) {
+            revert Rollup__NotInClaimPhase();
+        }
+
+        // Bond provider must have approved the contract to transfer the bond amount
+        testToken.transferFrom(_bondProvider, address(this), PROOF_COMMITMENT_BOND_AMOUNT);
```

Comment:
As earlier. This is vulnerable. But for conveying sentiment it might be fine, just REALLY don't want anyone to actually do this.

Reply:
Fixed! Flow indicates now that claim proof is to be performed by the bond provider.